pyrokinetics/gene.py

```diff
@@ -0,0 +1,370 @@
+        self.default_file_name = 'input.gene'
+
+    def read(self, pyro, data_file=None, template=False):
+        """
+        Reads GENE input file into a dictionary
+        """
+
+        if template and data_file is None:
+            data_file = self.base_template_file
+
+        gene = f90nml.read(data_file).todict()
+
+        pyro.initial_datafile = copy.copy(gene)
+
+        try: 
+            nl_flag = gene['general']['nonlinear']
+        
+            if nl_flag == '.F.':
+
+                pyro.linear = True
+            else: 
+                pyro.linear = False 
+        except KeyError:
+            pyro.linear = True 
+
+
+
+        pyro.gene_input = gene
+
+ #       # Loads pyro object with equilibrium data
+        if not template:
+            self.load_pyro(pyro)
+
+        # Load Pyro with numerics if they don't exist
+        if not hasattr(pyro, 'numerics'):
+            self.load_numerics(pyro, gene)
+
+    def load_pyro(self, pyro):
+        """
+        Loads GENE input into Pyro object
+        """
+
+        # Geometry
+        gene = pyro.gene_input
+
+        gene_eq = gene['geometry']['magn_geometry']
+
+        if gene_eq == 's_alpha': 
+            pyro.local_geometry = 'SAlpha'
+        elif gene_eq == 'miller': 
+            pyro.local_geometry = 'Miller'
+
+
+        #  Load GENE with local geometry
+        self.load_local_geometry(pyro, gene)
+
+        # Load GENE with species data
+        self.load_local_species(pyro, gene)
+
+
+ # Need species to set up beta_prime
+
+        if pyro.local_geometry_type == 'Miller':
+            if pyro.local_geometry['B0'] is not None:
+                pyro.local_geometry['beta_prime'] = - pyro.local_species['a_lp'] / pyro.local_geometry['B0'] ** 2
+            else:
+                pyro.local_geometry['beta_prime'] = 0.0
+        else:
+            raise NotImplementedError
+
+
+
+
+        # Load Pyro with numerics
+        self.load_numerics(pyro, gene)
+
+    def write(self, pyro, filename, directory='.'):
+        """
+        For a given pyro object write a GENE input file
+
+        """
+
+        if not os.path.exists(directory):
+            os.makedirs(directory)
+
+        path_to_file = os.path.join(directory, filename)
+
+        gene_input = pyro.gene_input
+
+        # Geometry data
+        if pyro.local_geometry_type == 'Miller':
+            miller = pyro.local_geometry
+
+            # Ensure Miller settings in input file
+            gene_input['geometry']['magn_geometry'] = 'miller'
+
+            # Reference B field
+            bref = miller['B0']
+
+            shat = miller['shat']
+            # Assign Miller values to input file
+            pyro_gene_miller = self.pyro_to_gene_miller()
+
+            # GENE uses definitions consistent with Miller. 
+            for key, val in pyro_gene_miller.items():
+                gene_input[val[0]][val[1]] = miller[key]
```

Comment:
Is it worth sticking to one kind of variable setting where possible. i.e. instead of ```miller['shat']```, we use ```miller.shat```?

Though there are cases like this loop when you iterate over keys where this is not really possible

Reply:
I think where possible it's probably a good idea.